handlers/secrets.go

```diff
@@ -28,30 +29,70 @@ const (
 
 // MakeSecretHandler makes a handler for Create/List/Delete/Update of
 //secrets in the Kubernetes API
-func MakeSecretHandler(namespace string, kube kubernetes.Interface) http.HandlerFunc {
+func MakeSecretHandler(defaultNamespace string, kube kubernetes.Interface) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		if r.Body != nil {
 			defer r.Body.Close()
 		}
 
-		client := kube.CoreV1().Secrets(namespace)
+		lookupNamespace, err, errCode := getLookupNamespace(defaultNamespace, r)
+		if err != nil {
+			http.Error(w, err.Error(), errCode)
+		}
+
+		client := kube.CoreV1().Secrets(lookupNamespace)
 
 		switch r.Method {
 		case http.MethodGet:
 			getSecretsHandler(client, w, r)
 		case http.MethodPost:
-			createSecretHandler(client, namespace, w, r)
+			createSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodPut:
-			replaceSecretHandler(client, namespace, w, r)
+			replaceSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodDelete:
-			deleteSecretHandler(client, namespace, w, r)
+			deleteSecretHandler(client, lookupNamespace, w, r)
 		default:
 			w.WriteHeader(http.StatusBadRequest)
 			return
 		}
 	}
 }
 
+// getLookupNamespace returns the namespace specified in any request type
+func getLookupNamespace(defaultNamespace string, r *http.Request) (string, error, int) {
+	lookupNamespace := defaultNamespace
+
+	if r.Method == http.MethodGet {
+		q := r.URL.Query()
+		namespace := q.Get("namespace")
+
+		if len(namespace) > 0 {
+			lookupNamespace = namespace
+		}
+	} else {
+		body, _ := ioutil.ReadAll(r.Body)
+		req := types.Secret{}
+		err := json.Unmarshal(body, &req)
+		if err != nil {
+			log.Printf("error while getting namespace: %s\n", err)
+			return "", fmt.Errorf("unable to unmarshal Secret request"), http.StatusBadRequest
+		}
+
+		// Reconstruct Body
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+		if req.Namespace != "" {
+			lookupNamespace = req.Namespace
+		}
+	}
+
+	if lookupNamespace == "kube-system" {
+		return "", fmt.Errorf("unable to list within the kube-system namespace"), http.StatusUnauthorized
```

Comment:
I think we can update this error message to something like .. `unable to manage secrets within the kube-system namespace`. which is related to what user is trying to do.

Reply:
Agreed and changed 💃

Reply:
Yes, but for every place it occurs, not just the one message.